src/test/pages/TeacherRequest.test.tsx

```diff
@@ -0,0 +1,204 @@
+import { TestUserRole, checkPageTitle, renderTestEnvironment } from "../utils";
+import { TeacherRequest } from "../../app/components/pages/TeacherRequest";
+import { RestHandler, rest } from "msw";
+import { API_PATH } from "../../app/services";
+import { screen } from "@testing-library/react";
+import { produce } from "immer";
+import { mockUser } from "../../mocks/data";
+import userEvent from "@testing-library/user-event";
+import * as actions from "../../app/state/actions";
+
+function checkDisabledAndCorrectValue(input: HTMLElement, expectedValue: string) {
+  expect(input).toHaveValue(expectedValue);
+  expect(input).toBeDisabled();
+}
+
+const upgradeAccountSpy = jest.spyOn(actions, "upgradeAccount");
+
+describe("TeacherRequest", () => {
+  const setupTest = ({
+    role,
+    extraEndpoints,
+    modifyUser,
+  }: {
+    role: TestUserRole;
+    extraEndpoints?: RestHandler<never>[];
+    modifyUser?: (u: typeof mockUser) => typeof mockUser;
+  }) => {
+    renderTestEnvironment({
+      role: role,
+      modifyUser: modifyUser,
+      PageComponent: TeacherRequest,
+      initialRouteEntries: ["/teacher_account_request"],
+      extraEndpoints: extraEndpoints,
+    });
+  };
+
+  it("renders without crashing", () => {
+    setupTest({ role: "STUDENT" });
+    checkPageTitle("Teacher account request");
+  });
+
+  it("displays a warning message if one is returned from the server", async () => {
+    const warningMessage = {
+      id: "teacher_registration_warning_message",
+      type: "isaacPageFragment",
+      canonicalSourceFile: "content/general_pages/fragments/teacher_registration_warning_message.json",
+      title: "Teacher registration",
+      encoding: "markdown",
+      children: [
+        {
+          type: "content",
+          encoding: "html",
+          children: [],
+          value: "<p>Test warning message</p>",
+          published: false,
+          tags: [],
+        },
+      ],
+    };
+    const extraEndpoints = [
+      rest.get(API_PATH + `/pages/fragments/teacher_registration_warning_message`, (req, res, ctx) => {
+        return res(ctx.status(200), ctx.json(warningMessage));
+      }),
+    ];
+    setupTest({ role: "STUDENT", extraEndpoints });
+    const warning = await screen.findByTestId("teacher-upgrade-warning");
+    expect(warning).toHaveTextContent("Test warning message");
+  });
+
+  it("does not display a warning message if none are returned from the server", () => {
+    const extraEndpoints = [
```

Comment:
I think this needs reworking, as it's a bit confusing at the moment. The path is called `teacher_registration_warning_message` but you're testing that there is no warning message. It also returns a 404, but I'm not sure how that would be a valid scenario.

Reply:
That's how it works actually 😀 Essentially there's a piece of content in the content repo that is currently marked as unpublished, so when fetched it returns a 404.  If/when the message is required to be shown, the content team just mark it as published, the front end can receive and display it then.  Not sure that's the best way of doing things, but that's how it's currently been designed.

Reply:
Ahhh, sorry, my bad. Thanks for clarifying! I thought that was the response from the Java API.